### PR TITLE
New storage: file to collection

### PR DESCRIPTION
**Pull request by wpferguson**
Created a new storage that exports an image then imports it back into the collection and groups it with the original.  Tags and ratings are copied from the original to the import.

### CODE REVIEW

File: contrib/export2collection.lua [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+dt.configuration.check_version(...,{3,0,0})
+
+-- Tell gettext where to find the .mo file translating messages for a particular domain
+gettext.bindtextdomain("export2collection",dt.configuration.config_dir.."/lua/locale/")
+
+local function _(msgid)
+    return gettext.dgettext("export2collection", msgid)
+end
+
+local function split_filepath(str)
```
Comment (boucman) [inline contrib/export2collection.lua]: you should start using dtutils, since it's on its way in :)

File: contrib/export2collection.lua [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+  -- for each of the image, exported image pairs
+  --   move the exported image into the directory with the original  
+  --   then import the image into the database which will group it with the original
+  --   and then copy over any tags other than darktable tags
+
+  for image,exported_image in pairs(image_table) do
+
+    local myimage_name = image.path .. "/" .. get_filename(exported_image)
+
+    while check_if_file_exists(myimage_name) do
```
Comment (boucman) [inline contrib/export2collection.lua]: this loop could be made into a utility function find_unique_file_name or something like that (in dtutils.file)

Comment (wpferguson) [inline contrib/export2collection.lua]: I agree.  Consider it done...

File: contrib/export2collection.lua (hunk 1 above)
Comment (wpferguson) [inline contrib/export2collection.lua]: I actually copied and pasted from dtutils.file for the new check_if.. functions.  ;)

**Comment by wpferguson** [comment]
There was a feature request in the mailing list and I looked at it and figured I could write that script in 30 minutes.  I took gimp.lua and ripped out the GIMP specific part and got this.  There are 2 use cases I can think of right now.  One is official/enfuse.lua requires ldr files, so this would be a way to create them then use the enfuse module.  The other is a user named Harry Durgin who has made several darktable videos.  His workflow is do a bunch of edits, then export to compress the history stack, then repeat.

Maybe, as Jeff Goldblum said in Jurassic Park, I should have asked "Should we?" instead of "Can we?".

**Comment by houz** [comment]
I wonder, shouldn't we rather add a signal that gets fired when an image got exported, so you can use the regular export storage and just import/group when an image got exported? Or is this something that is only intended to be used once in a while?

**Comment by boucman** [comment]
lua can already add storages (if I understand houz correctly) but the point is that we are talking about tools that do image=>external-tool=>reimport.

the general, vague, consensus is that such a tool shouldn't be a strorage (from a UI point of view) but something else... it can't really be a iop either (even by stretching the definition) because it can involve multiple images (enfuse)

I'm not sure how to call such tools, nor how to handle them in the UI. Right now we are (poorly) reproducing the export-lib UI in lua in order to separate them from storage. Maybe we could have a separate export-like UI for that (on the C side, I mean) I don't know...

@wpferguson : I am a bit confuse by the "export to compress the history stack" step you describe... what do you mean ? is that a workaround for a DT bug ?

**Comment by houz** [comment]
I know that we can add Lua storages. What I meant was a new signal, similar to the intermediate-export-image one, which a script can register for, getting the filename of the exported image + image object of the original + the storage used. Then the script could check if the storage is "disk" and import the exported image and group it with the original.

**Comment by boucman** [comment]
Ok, so right now there is no real point in doing this, i'll close the PR

**Comment by wpferguson** [comment]
@boucman here is a link to one of harry durgin's videos - https://www.youtube.com/watch?v=QbQA2P2EM1w.  At 26 minutes he exports his raw to a tif and brings it back in.  My understanding of what he is doing is applying all the changes so that he doesn't have to wait while the image processes to update the display when he changes something.  It may also be that he doesn't know how to compress the history stack, so he thinks darktable is doing more processing than it actually is.

**Comment by houz** [comment]
He does that in every video for two reasons:

- to make darktable faster as he tends to pile up duplicates of the modules
- to battle the fixed pipe order. By using intermediate exports he can use an early module on the result of a late one.

**Pull request closed without merging**